### PR TITLE
Core/Pet: Make pet load asynchronous

### PR DESCRIPTION
**Pull request by ariel-**
**Changes proposed:**

-  Moved queries to async, uses callbacks to finish job after load

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [x] master

**Issues addressed:** 

**Tests performed:** Built, tested in-game during valgrind/helgrind run

**Known issues and TODO list:** None AFAIK

### CODE REVIEW

File: src/server/game/Entities/Pet/Pet.cpp [hunk 1]
```diff
@@ -135,18 +144,67 @@ bool Pet::LoadPetFromDB(Player* owner, uint32 petEntry, uint32 petnumber, bool c
...
-    result = CharacterDatabase.Query(stmt);
+    WorldSession* session = owner->GetSession();
+    Pet* pet = this;
+    session->GetQueryProcessor().AddQuery(CharacterDatabase.AsyncQuery(stmt)
+        .WithChainingPreparedCallback([pet, session, callback, current](QueryCallback& queryCallback, PreparedQueryResult result)
+    {
+        Player* player = session->GetPlayer();
+        if (!player || !player->IsInWorld())
+        {
+            delete pet;
```
Comment (jackpoz) [inline src/server/game/Entities/Pet/Pet.cpp]: it looks quite dangerous to me to delete this, changing the usual common handling of LoadFromDB() result.

Comment (ariel-) [inline src/server/game/Entities/Pet/Pet.cpp]: Moved from

https://github.com/TrinityCore/TrinityCore/pull/19747/files#diff-5589ee47b3a7ad220cff105a8fe7eac3L18406
https://github.com/TrinityCore/TrinityCore/pull/19747/files#diff-5589ee47b3a7ad220cff105a8fe7eac3L25285

Comment (jackpoz) [inline src/server/game/Entities/Pet/Pet.cpp]: it's not moved, it's a different code behavior, from deleting the pointer returned by new to delete this (obviously with "delete this" I mean C++ code "delete this;")

Comment (ariel-) [inline src/server/game/Entities/Pet/Pet.cpp]: What exactly makes "delete this" dangerous if you return immediately after it? I've already seen this pattern in the core. Is there any alternate solution?

Comment (joschiwald) [inline src/server/game/Entities/Pet/Pet.cpp]: pets can't exist without player owner

Comment (jackpoz) [inline src/server/game/Entities/Pet/Pet.cpp]: > Is there any alternate solution?

The old behavior is the most common one, with the owner of new Pet() having to delete it when needed.
The caller has no way to know that the pointer is no longer valid, that's what makes it dangerous imo.

Comment (ariel-) [inline src/server/game/Entities/Pet/Pet.cpp]: The caller won't get a pointer in this case, the callback function is called with a bool and a Pet*.

When the bool is true, the loading succeeded, if it's false it failed, In either case the pointer is a valid one, and allows for error handling.

Comment (ccrs) [inline src/server/game/Entities/Pet/Pet.cpp]: easier than discussing if its dangerous or not, just test the cases in which it could be dangerous 💃

Comment (tripleslash) [inline src/server/game/Entities/Pet/Pet.cpp]: Smart Pointers boys

Comment (Faq) [reply src/server/game/Entities/Pet/Pet.cpp]: Just wondering who won in this argument:
@jackpoz vs @ariel-  or @tripleslash suggestion

**Comment by tkrokli** [comment]
If I made any mistake in my conflict solving, feel free to improve it if you know how it should be.

**Comment by jackpoz** [comment]
@ariel- could you take a look at https://travis-ci.org/TrinityCore/TrinityCore/builds/306879619#L1728 so then we can merge the PR ?

**Comment by jackpoz** [comment]
@tkrokli which files conflicted ? your merge doesn't report any conflicted file. Also the build broke and I think it's because of the merge.
Maybe it's better to leave the merge conflicts instead since now it looks worse.

**Comment by jackpoz** [comment]
Yeah, looking at the merge it's just easier to throw it away and create a new branch with only the PR commit.

**Comment by jackpoz** [comment]
Fixed the merge conflicts for now, anyone up to do some ingame tests ?

**Comment by tkrokli** [comment]
I would definitely like to test it, preferably tomorrow evening at the earliest (and during the coming week). My main experience is from the hunter class. Could you say a few words about what to look for, except for obvious negative behaviour (if any)?

**Comment by ccrs** [comment]
I remember ariel telling me this had something big missing, should not be merged

**Comment by ariel-** [comment]
There were a little more sync queries I wish to be gone. Also this:

> it looks quite dangerous to me to delete this, changing the usual common handling of LoadFromDB() result.

I shall have a lot of fun revisiting the code after this time.

**Comment by jackpoz** [comment]
well you explained the "delete this" code was all ok so that point is fine. up to you if you want to edit the other sync queries another time or wait for the merge.

**Comment by joschiwald** [comment]
i would close this pr in favor to #21016, it would create more and more problems to have 2 different implementation in master and 3.3.5 branch

**Comment by ariel-** [comment]
#21016 does not make pet load async, for me is unrelated. But I can put this on hold till 21016 is merged, then adapt the code

**Comment by jackpoz** [comment]
I'd prefer to get this PR merged since it changes so little compared to the other PR (which is for master branch)

**Comment by joschiwald** [comment]
the other pr makes this pr obsolete once the loading part is fixed correctly and it makes async loading very easy

**Pull request closed without merging**